src/main/docker/bootstrap.sh

```diff
@@ -0,0 +1,52 @@
+#!/bin/bash
+
+cd /opt/Artemis
+
+if [ -z "$(ls -A config)" ]; then
+   echo "Config is Empty .. copying default ones .."
+   cp -n -a /defaults/Artemis/. config/
+else
+   echo "Config is not empty .. not copying default configs .."
+fi
+
+# Create sync file that indicates waiting for services
+mkdir -p sync
+touch sync/waiting-backend
+
+# Ensure at least the directories are owned by artemis. "-R" takes too long
+chown artemis:artemis config data sync
+
+# Allow waiting for other services
+if [ -n "${WAIT_FOR}" ]; then
+  hosts_ports=$(echo $WAIT_FOR | tr "," "\n" | tr -d "\"")
+  for host_port in $hosts_ports
+  do
+    until [[ "$(curl -s -o /dev/null -L -w ''%{http_code}'' http://$host_port)" == "200" ]]
+    do
+      echo "Waiting for $host_port"
+      sleep 5
+    done
+  done
+fi
+
+# Release indicator of startup after 90 seconds (may be used by nginx proxy to determine startup)
+bash -c "sleep 90 && rm -f sync/waiting-backend" &
+
+echo "Starting application..."
+exec gosu artemis java \
```

Comment:
Why are you wrapping the execution with `gosu`? The Docker internal `USER` directive would also work here right?

Reply:
No. To ensure the correct folder permissions, root access is needed (see bootstrap chown part)

Reply:
That also the cause why you need a script instead of writing the whole command in a CMD or ENTRYPOINT section (would be too much for such a secion :)).

Reply:
Ok .. you get an exception (and it's not clear what that breaks)

```
2022-04-01 10:22:09.671 ERROR 10 --- [           main] d.t.i.w.a.service.user.UserService       : An error occurred after application startup when creating or updating the admin user or in the LDAP search

de.tum.in.www1.artemis.service.connectors.gitlab.GitLabException: Error while trying to update user in GitLab: User{login='artemis', firstName='Artemis', lastName='Administrator', email='artemis-praktomat-test@cs.kit.edu', imageUrl='null', activated='true', langKey='en', activationKey='null'}
        at de.tum.in.www1.artemis.service.connectors.gitlab.GitLabUserManagementService.updateVcsUser(GitLabUserManagementService.java:72)
        at de.tum.in.www1.artemis.service.user.UserService.lambda$updateUserInConnectorsAndAuthProvider$10(UserService.java:389)
        at java.base/java.util.Optional.ifPresent(Optional.java:178)
        at de.tum.in.www1.artemis.service.user.UserService.updateUserInConnectorsAndAuthProvider(UserService.java:389)
        at de.tum.in.www1.artemis.service.user.UserService.applicationReady(UserService.java:120)
        at de.tum.in.www1.artemis.service.user.UserService$$FastClassBySpringCGLIB$$1cd375b3.invoke(<generated>)
        at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
        at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:689)
        at de.tum.in.www1.artemis.service.user.UserService$$EnhancerBySpringCGLIB$$bd31da70.applicationReady(<generated>)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.springframework.context.event.ApplicationListenerMethodAdapter.doInvoke(ApplicationListenerMethodAdapter.java:344)
        at org.springframework.context.event.ApplicationListenerMethodAdapter.processEvent(ApplicationListenerMethodAdapter.java:229)
        at org.springframework.context.event.ApplicationListenerMethodAdapter.onApplicationEvent(ApplicationListenerMethodAdapter.java:166)
        at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:176)
        at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:169)
        at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:143)
        at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:421)
        at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:378)
        at org.springframework.boot.context.event.EventPublishingRunListener.ready(EventPublishingRunListener.java:114)
        at org.springframework.boot.SpringApplicationRunListeners.lambda$ready$6(SpringApplicationRunListeners.java:82)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
        at org.springframework.boot.SpringApplicationRunListeners.doWithListeners(SpringApplicationRunListeners.java:120)
        at org.springframework.boot.SpringApplicationRunListeners.doWithListeners(SpringApplicationRunListeners.java:114)
        at org.springframework.boot.SpringApplicationRunListeners.ready(SpringApplicationRunListeners.java:82)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:318)
        at de.tum.in.www1.artemis.ArtemisApp.main(ArtemisApp.java:62)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:49)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:108)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:58)
        at org.springframework.boot.loader.WarLauncher.main(WarLauncher.java:59)
Caused by: org.gitlab4j.api.GitLabApiException: <html>
<head><title>502 Bad Gateway</title></head>
<body>
<center><h1>502 Bad Gateway</h1></center>
<hr><center>nginx/1.21.6</center>
</body>
</html>

        at org.gitlab4j.api.AbstractApi.validate(AbstractApi.java:642)
        at org.gitlab4j.api.AbstractApi.get(AbstractApi.java:214)
        at org.gitlab4j.api.UserApi.getUser(UserApi.java:317)
        at de.tum.in.www1.artemis.service.connectors.gitlab.GitLabUserManagementService.updateBasicUserInformation(GitLabUserManagementService.java:88)
        at de.tum.in.www1.artemis.service.connectors.gitlab.GitLabUserManagementService.updateVcsUser(GitLabUserManagementService.java:58)
        ... 36 common frames omitted

2022-04-01 10:22:09.721 DEBUG 10 --- [cached.thread-5] d.t.i.w.a.s.s.quiz.QuizScheduleService   : Process cached quiz submissions
2022-04-01 10:22:09.734  INFO 10 --- [           main] d.t.i.w.a.s.s.quiz.QuizScheduleService   : QuizScheduleService was started to run repeatedly with 5.0 second delay.
2022-04-01 10:22:09.754  INFO 10 --- [           main] d.t.i.w.a.s.s.quiz.QuizScheduleService   : Found 0 quiz exercises with planned start in the future
2022-04-01 10:22:09.758  INFO 10 --- [           main] de.tum.in.www1.artemis.ArtemisApp        :
----------------------------------------------------------
        'Artemis' is running! Access URLs:
        Local:     http://localhost:8080/
        External:  http://10.1.0.5:8080/
        Profiles:  [prod, jenkins, gitlab, artemis, scheduling, saml2]
        Version:   5.7.0
----------------------------------------------------------
```

Do you think that this does not cause problems ? Artemis is running but Exceptions at startup are suspicious :)
(As far as I remember, in the past the whole artemis process crashed :) )

Reply:
(I think this is the wrong thread 😬 anyway) 

I asked @sleiss (please correct me if I understood that wrong) about it: 

Artemis syncs the `internal_admin` account (and password) to the gitlab instance on startup. This is a conveniency feature since an Artemis administrator may use the artemis_admin account to login to gilab or start a coding exercise. This sync only has to happen once at the initial Artemis setup (and every time you change the password of the artemis internal admin account). 
A quick workaround for the password update is to restart the Artemis container after all other containers are up and running. 
So yes - This exception can be ignored (in most cases ;) )

Reply:
I've hidden the comments to keep the threads clear. Sorry for posting to the wrong thread :D